design/vela-core/definition-versioning.md

```diff
@@ -0,0 +1,296 @@
+  latest when no Component Revision is specified. 
+  > While we don't ideally want Application developers to bother with such details, there are use cases where
+  > an automatic upgrade to the latest Component version is not desired.
+
+
+### Goals
+
+- Support Component versioning with Semantic Versions.
+- Allow pinning, specific and non-specific versions of a Component in the
+  Kubevela Application.
+
+### Non-Goals
+
+- Support for version range in Application. For eg. "type: my-component@>1.2.0"
+
+## Acceptance Criteria
+
+**User Story: Component version specification**
+
+>**AS A** Component author\
+>**I SHOULD** be able to publish every version of my Component with the Semantic Versioning scheme\
+>**SO THAT** an Application developer can use a specific version of the Component.
+
+**BDD Acceptance Criteria**
+
+>**GIVEN** a Component spec\
+>**AND** a version field of the spec set to V\
+>**WHEN** the Component is applied to the KubeVela\
+>**THEN** it should be listed as one of the many versions in the DefinitionRevision list
+
+**User Story: Application Component version specification**
+
+>**AS AN** Application developer\
+>**I SHOULD** be able to specify a version (complete or partial) for every Component used\
+>**SO THAT** I can control which version are deployed.
+
+**BDD Acceptance Criteria**
+
+>Scenario: Use the version specified in the Application manifest when deploying the service\
+>**GIVEN** a Component A with version 1.2.3\
+>**AND** a Component B with version 4.5.6\
+>**AND** an Application composed of A 1.2.2 and B 4.4.2\
+>**WHEN** the Application is deployed\
+>**THEN** it uses Component A 1.2.2 and B 4.4.2
+
+>**Variant:** Use the latest version for the part of the semVer that is not specified.\
+>**AND** an Application composed of A 1.2 and B 4\
+>**WHEN** the Application is deployed\
+>**THEN** it uses Component A 1.2.3 and B 4.5.6
+
+## Current Implementation
+
+Currently, Kubevela has some support for controlling Definition versions based on K8s annotations and DefinitionRevisions. The annotation `definitionrevision.oam.dev/name` can be used to version the ComponentDefinition. For example if the following annotation is added to a ComponentDefinition, it produces a new DefinitionRevision and names the ComponentDefinition as `component-name-v4.4` .
+
+> definitionrevision.oam.dev/name: "4.4"
+
+![version](./kubevela-version.png)
+
+This Component can then be referred in the Application as either of the following:
+
+>"component-name@v4.4" - `NamedDefinitionRevision`\
+>"component-name@v3"   - `DefinitionRevision`
+
+This versioning scheme, although convenient, has the following issues:
+
+- Since, there are more than one ways to reference a Component (`NamedDefinitionRevision` or `DefinitionRevision`), it proves difficult to maintain version consistency across clusters.
```

Comment:
For app developers, I don't think they need to notice the differences, they only need to focus on X-Definitions. What about adding a `version` field in component definition?

Reply:
Agreed. But in code we will have to figure out a way to distinguish between components created with `name` and those created with `revision` to apply auto-upgrade only to the latter.